From 44254e4cc147a93caa55e1d01bfd52d7887eceb8 Mon Sep 17 00:00:00 2001
From: Doze-off <Doze@off.com>
Date: Wed, 16 Jul 2025 05:50:01 -0300
Subject: [PATCH] Fix: Prevent NullPointerException in BpfNetMaps during UID
 removal

This patch adds null checks and exception handling when removing UIDs
from sLocalNetBlockedUidMap in BpfNetMaps and PermissionMonitor.
It prevents potential system_server crashes during boot or app
install/uninstall events on devices where the BPF UID map may be null.

This change improves system stability and avoids random reboots
also tested on custom roms, with old kernels 4.9 4.14
---
 service/src/com/android/server/BpfNetMaps.java                | 4 ++++
 .../com/android/server/connectivity/PermissionMonitor.java    | 2 ++
 2 files changed, 6 insertions(+)

diff --git a/service/src/com/android/server/BpfNetMaps.java b/service/src/com/android/server/BpfNetMaps.java
index 4233268..3de12b0 100644
--- a/service/src/com/android/server/BpfNetMaps.java
+++ b/service/src/com/android/server/BpfNetMaps.java
@@ -1935,6 +1935,10 @@ public class BpfNetMaps {
     @RequiresApi(Build.VERSION_CODES.CUR_DEVELOPMENT)
     public void removeUidFromLocalNetBlockMap(final int uid) {
         throwIfPre25Q2("removeUidFromLocalNetBlockMap is not available on pre-B devices");
+        if (sLocalNetBlockedUidMap == null) {
+            Log.w(TAG, "sLocalNetBlockedUidMap is null, skipping delete for uid: " + uid);
+            return;
+        }
         try {
             sLocalNetBlockedUidMap.deleteEntry(new U32(uid));
         } catch (ErrnoException e) {
diff --git a/service/src/com/android/server/connectivity/PermissionMonitor.java b/service/src/com/android/server/connectivity/PermissionMonitor.java
index 0f88270..aa8a505 100755
--- a/service/src/com/android/server/connectivity/PermissionMonitor.java
+++ b/service/src/com/android/server/connectivity/PermissionMonitor.java
@@ -1335,6 +1335,7 @@ public class PermissionMonitor {
         // mUidToNetworkPerm to check if the package can bypass VPN.
         updateVpnUid(uid, false /* add */);
         updateLockdownUid(uid, false /* add */);
+
         // If the package has been removed from all users on the device, clear it form mAllApps.
         if (mPackageManager.getNameForUid(uid) == null) {
             mAllApps.remove(appId);
@@ -1374,6 +1375,7 @@ public class PermissionMonitor {
         }
     }
 
+
     private static int getNetdPermissionMask(String[] requestedPermissions,
                                              int[] requestedPermissionsFlags) {
         int permissions = PERMISSION_NONE;
